gremlin-server/src/main/java/org/apache/tinkerpop/gremlin/server/handler/AbstractSession.java

```diff
@@ -0,0 +1,819 @@
+     */
+    protected Map<String, Object> generateResponseMetaData(final SessionTask sessionTask,
+                                                           final ResponseStatusCode code, final Iterator<?> itty) {
+        return Collections.emptyMap();
+    }
+
+    protected Frame makeFrame(final SessionTask sessionTask, final List<Object> aggregate,
+                              final ResponseStatusCode code, final Iterator<?> itty) throws Exception {
+        final RequestMessage msg = sessionTask.getRequestMessage();
+        final ChannelHandlerContext nettyContext = sessionTask.getChannelHandlerContext();
+        final MessageSerializer serializer = nettyContext.channel().attr(StateKey.SERIALIZER).get();
+        final boolean useBinary = nettyContext.channel().attr(StateKey.USE_BINARY).get();
+
+        final Map<String, Object> responseMetaData = generateResponseMetaData(sessionTask, code, itty);
+        final Map<String, Object> statusAttributes = generateStatusAttributes(sessionTask, code, itty);
+        try {
+            if (useBinary) {
+                return new Frame(serializer.serializeResponseAsBinary(ResponseMessage.build(msg)
+                        .code(code)
+                        .statusAttributes(statusAttributes)
+                        .responseMetaData(responseMetaData)
+                        .result(aggregate).create(), nettyContext.alloc()));
+            } else {
+                // the expectation is that the GremlinTextRequestDecoder will have placed a MessageTextSerializer
+                // instance on the channel.
+                final MessageTextSerializer textSerializer = (MessageTextSerializer) serializer;
+                return new Frame(textSerializer.serializeResponseAsString(ResponseMessage.build(msg)
+                        .code(code)
+                        .statusAttributes(statusAttributes)
+                        .responseMetaData(responseMetaData)
+                        .result(aggregate).create()));
+            }
+        } catch (Exception ex) {
+            logger.warn("The result [{}] in the request {} could not be serialized and returned.", aggregate, msg.getRequestId(), ex);
+            final String errorMessage = String.format("Error during serialization: %s", ExceptionHelper.getMessageFromExceptionOrCause(ex));
+            final ResponseMessage error = ResponseMessage.build(msg.getRequestId())
+                    .statusMessage(errorMessage)
+                    .statusAttributeException(ex)
+                    .code(ResponseStatusCode.SERVER_ERROR_SERIALIZATION).create();
+            sessionTask.writeAndFlush(error);
+            throw ex;
+        }
+    }
+
+    /**
+     * Called right before a transaction starts within {@link #run()}.
+     */
+    protected void startTransaction(final SessionTask sessionTask) {
+        // check if transactions are open and rollback first to ensure a fresh start.
+        graphManager.rollbackAll();
```

Comment:
if transactions are open, it means we have a bug in our code where we are failing to properly close transactions. please add a warn log here. I am actually in favor of throwing an exception and not processing this request at all because an invalid transaction state is a serious bug which can impact correctness of data.

Reply:
agreed to throw an exception. we've always been sorta relaxed about handling open transactions and that was probably a mistake. quietly rolling back just hides the problem. good idea.